### ISSUE TITLE
Create single source for deriving FD command sizes

### ISSUE THREAD

**Issue by abhullar-tt** [tenstorrent/tt-metal#16149]
Currently we use DeviceCommand to construct FD commands with APIs but do a lot of cmd size calculations in the caller (command_queue.cpp). This has been causing a lot of alignment related bugs (exposed by BH having different pcie alignment than WH)

We should expose helpers so we don't recalculate sizes twice.

**Comment by abhullar-tt** [comment]
For e.g. when we look at `EnqueueRecordEventCommand::process` we do the following:

1. We calculate the size of the command that will be written to the issue queue. Host writes commands into the issue queue (and owns the write ptr) and Fast Dispatch kernels on device read from issue queue and act on the commands (kernel owns read ptr). The issue queue is a region within a Hugepage
```
uint32_t cmd_sequence_sizeB =
    hal.get_alignment(HalMemType::HOST) * num_worker_counters +  // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WAIT
    packed_write_sizeB +  // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WRITE_PACKED + unicast subcmds + event
                          // payload
    align(
        sizeof(CQPrefetchCmd) + sizeof(CQDispatchCmd) + dispatch_constants::EVENT_PADDED_SIZE,
        pcie_alignment);  // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WRITE_LINEAR_HOST + event ID
```
2. Reserve space for command in the issue queue (wrapping the write ptr if necessary)
```
void* cmd_region = this->manager.issue_queue_reserve(cmd_sequence_sizeB, this->command_queue_id);
```
3.  Create a `DeviceCommand` object that allows commands to be written directly to the issue queue. For performance reasons we want to add commands directly into the issue queue rather than constructed in another buffer then copying it to the issue queue.
```
HugepageDeviceCommand command_sequence(cmd_region, cmd_sequence_sizeB);
```
4. Add the Fast Dispatch commands
```
command_sequence.add_dispatch_wait(...)
...
command_sequence.add_dispatch_write_packed<CQDispatchWritePackedUnicastSubCmd>(...)
...
command_sequence.add_dispatch_write_host<true>(...)
...
```
5. Update issue queue write pointer and notify Fast Dispatch kernel of updated write
```
this->manager.issue_queue_push_back(cmd_sequence_sizeB, this->command_queue_id);

this->manager.fetch_queue_reserve_back(this->command_queue_id);
this->manager.fetch_queue_write(cmd_sequence_sizeB, this->command_queue_id);
```

Not having APIs for calculating size is error prone because if a miscalculation could result differences between calculated size and what command is actually constructed. This has lead to Fast Dispatch kernels reading out garbage commands. 

Instead of step 1 manually calculating the size of commands we could have static apis in DeviceCommand that return the size of the command based on user provided parameters. The caller knows how the commands need to be constructed so they would have this info to supply to the apis. 

@tt-aho and @jbaumanTT wondering if you guys have more thoughts on this.

**Comment by jbaumanTT** [comment]
Yeah, this is a wart I've been wanting to fix. It might make sense to have a stateful API so we could mirror the DeviceCommandSequence code more closely and be able to handle alignment issues if we chain allocators in more complicated ways (e.g. add_dispatch_write_packed_large and then add_prefetch_relay_paged_packed).

**Comment by jbaumanTT** [comment]
Also, we should assert that the amount of command bytes written equals cmd_sequence_sizeB, which would catch most mismatches.
